docs/designs/keras_callback.md

```diff
@@ -0,0 +1,208 @@
+        return 0.0002
+    else:
+        return 0.0001
+
+# Process the prediction outputs for each batch
+class PredictionOutputsProcessor(BasePredictionOutputsProcessor):
+    ...
+```
+There will be multiple definition APIs for users to define different behaviors of model and the interfaces of APIs may be different. It is more convenient for users to define those behaviors using `tf.keras.callbacks.Callback`. 
+
+Some use cases we observed that we want to support are:
+* Case 1: Callback similar to `PredictionOutputsProcessor` that is executed after prediction outputs are made.
+* Case 2: Callback to modulate learning rate as we are currently doing in `LearningRateScheduler`.
+* Case 3: Callback to checkpoint model weights during training.
+* Case 4: Callback to write additional summaries to TensorBoard after each evaluation job completes.
+* Case 5: Callback to perform early stopping when certain conditions are met. For example, the metrics are met after an evaluation job.
+* Case 6: Callback to upload model to remote storage after completing training.
+
+
+To implement callbacks, we need to solve the following problems:
+1. How to define callbacks for the model in ElasticDL.
+2. How to call the methods of callbacks in ElasticDL.
+
+## Define Callbacks for the Model in ElasticDL
+
+In the model definition file, users can add `callbacks` API to define callbacks like:
+```python
+def custom_model():
+    ...
+
+def callbacks():
+    early_stop_callback = EarlyStopCallback()
+    learning_rate_scheduler = LearningRateScheduler()
+    return [early_stop_callback, learning_rate_scheduler]
+
+
+class EarlyStopCallback(tf.keras.callbacks.Callback):
+    def on_batch_end(self, batch, logs=None):
+        ...
+
+class LearningRateScheduler(tf.keras.callbacks.Callback):
+    def on_batch_start(self, batch, logs=None):
+        ...
+```
+
+## Call the Methods of Callback in ElasticDL
+
+We may define several callbacks for the model in a job. Tensorflow creates a container `CallbackList` to wrap the callbacks to conveniently call the methods in callbacks. We can call the methods in callbacks by calling the methods in `CallbackList`. For example:
+```python
+class CallbackList():
+    def on_batch_end(self, batch, logs=None):
+        for callback in self.callback_list:
+            callback.on_batch_end(bach, logs)
+```
+For detail, we can view the [source code](https://github.com/tensorflow/tensorflow/blob/cf7fcf164c9846502b21cebb7d3d5ccf6cb626e8/tensorflow/python/keras/callbacks.py#L189-L196
+) of `CallbackList` in Tensorflow.
+
+So, we can also use CallbackList to wrap the callbacks in ElasticDL.
+```python
+from tensorflow.python.keras.callbacks import CallbackList
+callbacks = callbacks()
+callback_list = CallbackList(callbacks)
+callback_list.on_batch_end(batch)
+```
+
+
+### Call the Methods of Callback using ParameterServerStrategy
+
+Using ParameterServerStrategy in ElasticDL, we will launch instances as master, worker and parameter server. 
```

Comment:
The Callback interface contains multiple methods.
How do we decide which methods execute on master or worker? From the method name?

Reply:
Not from the method name but call frequency. The worker call methods per batch and the master call methods per epoch and job.